### PR TITLE
Add mediation fees ADR

### PR DESCRIPTION
**Pull request by karlb**
I've changed the proportional fee to a float value again, since we won't be 100% accurate anyway due to the interpolation of the imbalance penalty function.

Did I forget anything? Do any points need more clarification? Did I assume consensus on something we didn't agree on?

Closes https://github.com/raiden-network/raiden-services/issues/341.

### CODE REVIEW

File: adr/003-mediation-fees.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+1. A fixed fee + a proportional fee (flat + prop * amount)
+2. A fixed fee + a proportional fee + an imbalance fee based on a fixed imbalance measure
+3. A fixed fee + a proportional fee + submitting points of a function describing the imbalance fee
+4. Submitting points of a single function (dependent on the payment amount) which includes all fees
+
+## Decision Outcome
+
+Chosen option: **Option 3**, but with some important additions (see below).
+
+Option 1 would be sufficient for the short very term, but does not incentivize rebalancing. Option 2 does not work because no objective balance measure could be found that is valid across all business models / user interests. Option 4 would work but require sending a changed fee schedule after every transfer or mediation. Since this information is broadcast, that's a non-negligible amount of traffic.
```
Comment (palango) [inline adr/003-mediation-fees.md]: Maybe make this a list for each option?

File: adr/003-mediation-fees.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+### Positive Consequences
+
```
Comment (palango) [inline adr/003-mediation-fees.md]: I'd add that balance fees can also be disabled.

File: adr/003-mediation-fees.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+* Fully exhausting a channel can be punished more strongly than just moving into the wrong direction
+* Fee schedules only have to be changed when the node's strategy changes or the on-chain capacity changes via `deposit` or `withdraw`
+
+### Negative Consequences
+
+* Comparatively complex
+
+## Detailed analysis
+
+One important detail is that fees are not only calculated for the channel over which the mediation payment is sent, but also for the channel over which the payment is received. Without this, a node could not incentivize receiving through a certain channel, which is just as important for rebalancing as sending through the right channel. So the total mediation fee for a node is
```
Comment (palango) [inline adr/003-mediation-fees.md]: Maybe the example we had at the whiteboard would help to understand why this is desirable.

File: adr/003-mediation-fees.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+`MF_total = MF_in_channel + MF_out_channel`
+
+where for each channel
+
+`MF_channel = flat_fee + proportional_fee * amount + IP(C_before) - IP(C_after)`
```
Comment (palango) [inline adr/003-mediation-fees.md]: I thought we said that flat and proportional fees are only applicable for the outgoing channel.

File: adr/003-mediation-fees.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+where `IP` is the "Imbalance Penalty" function and `C_before/after` are the channel capacities before and after the payment. The channel capacity is the node's free capacity as reported to the PFS with the UpdatePFS messages.
+
+The IP function is a function that describes how much a node is willing to pay to move away from an undesirable channel capacity. If a node prefers to have a channel capacity of 5 while the total capacity of that channel is 10 (so that it could mediate up to 5 tokens in both directions) the IP function might look like
```
Comment (palango) [inline adr/003-mediation-fees.md]: We need to clarify if IP is relative or absolute values, and if it's token or token weis.

File: adr/003-mediation-fees.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+In this model, the mediation fees are split into three parts:
+
+1. The *flat fee*. This is meant to represent the base cost of handling a payment, so that even small payment amounts can be profitable to mediate.
+2. The *proportional fee*. Large payments lock up tokens while the payment is in progress. To compensate this, the operator should receive larger fees for large payments
+3. The *imbalance fee*. The direction of payments and the specific channels used are relevant to the operator. Some payments might exhaust the channel capacity and prevent future profitable mediations while others could facilitate future mediations. Due to this, the operator should be compensated (or willing to compensate the initiator, respectively). As a side effect, the whole network will get into a more balanced state.
+
+These three parts roughly map to operating costs, capital costs and opportunity costs.
+
+The flexible shape of the IP function allows to encode many different operator intentions. If I use a channel solely to pay, having more free capacity is always desired and any capacity gained by mediating in the reverse direction is welcome. In that case, my IP function could look like
```
Comment (palango) [inline adr/003-mediation-fees.md]: Nice examples! 👍

File: adr/003-mediation-fees.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+The message format still has to be defined in detail, but the payload will look roughly like this:
+
+```json
+{
+    'flat': 10,
+    // we have to deal with float values anyway, due to the interpolation
+    'proportional': 0.0001,  // factor of transfer amount
+    'imbalance_penalty': [
+        (0, 1000),
```
Comment (palango) [inline adr/003-mediation-fees.md]: These should be lists, there are no tuples in json.

**Comment by palango** [comment]
@hackaugusto Please give this a look.

File: adr/003-mediation-fees.md (hunk 4 above)
Comment (karlb) [reply adr/003-mediation-fees.md]: We could have the flat and proportional fee either
1. set per node
2. set per channel, used only for outgoing channel
3. set per channel, used for both incoming and outgoing channel

Option 2 feels like slightly awkward compromise to me. Option 1 would introduce fee information per node, which we wouldn't have with 2 or 3.

File: adr/003-mediation-fees.md (hunk 2 above)
Comment (karlb) [reply adr/003-mediation-fees.md]: Under "Message Format" I write

> The three mediation fee components are sent separately. Each component is optional and will be assumed to be zero (no cost) if not included in the message.

Do you expect a different statement, or do you just want this information in another place of the document?

File: adr/003-mediation-fees.md (hunk 5 above)
Comment (karlb) [reply adr/003-mediation-fees.md]: I'll add
> The IP function is a function that describes how much a node is willing to pay The IP function is a function that describes how much a node is willing to pay **(as absolute amount or RDN-wei)** to move ...

RDN-wei is not something we've used before, but at least it's not ambiguous.

Comment (palango) [reply adr/003-mediation-fees.md]: I think we call it Rei, for RDN wei

File: adr/003-mediation-fees.md (hunk 2 above)
Comment (palango) [reply adr/003-mediation-fees.md]: I would just state it as an advantage of the solution.

File: adr/003-mediation-fees.md (hunk 5 above)
Comment (karlb) [reply adr/003-mediation-fees.md]: I just noticed after commenting that we're not talking about RDN and edited my question.

File: adr/003-mediation-fees.md (hunk 2 above)
Comment (karlb) [reply adr/003-mediation-fees.md]: I've added

> Fee components are optional. A node only using flat fees does not have to submit complicated functions.

File: adr/003-mediation-fees.md (hunk 3 above)
Comment (karlb) [reply adr/003-mediation-fees.md]: Added a section for the example.

File: adr/003-mediation-fees.md [hunk 8]
```diff
@@ -74,6 +77,24 @@ If the node currently has a capacity of 6 and is asked to mediate a payment of 3
 
 Only mediating nodes demand mediation fees. The initiator and target could theoretically benefit if their IP was taken into account during routing, but this aspect is left out for now to avoid additional complexity.
 
+### Why apply fees for the incoming channel?
+
+At a first glance, calculating fees for the incoming channel does not make any sense, since the mediating node itself only sends money over the outgoing channel. But taking part in the mediation will modify both channels channels, so the choice of the incoming channel is relevant to the mediating node. To demonstrate that you can't reliably influence the balancing of your channel if you only apply fees to the outgoing channel, let's look at one example:
+
+A mediating node has three channels. The channel to C is nearly exhausted, while the other channels have a free capacity of 9 tokens left (of 10 tokens which are deposited in that channel). This node wants to have
+a similar amount of free capacity for all channel, so that it has more mediation opportunities. To achieve this, it must receive tokens over channel C.
```
Comment (palango) [inline adr/003-mediation-fees.md]: for all channel**s**

File: adr/003-mediation-fees.md [hunk 9]
```diff
@@ -74,6 +77,24 @@ If the node currently has a capacity of 6 and is asked to mediate a payment of 3
...
+++: desirable transfer (point of view of the mediating node)<br/>
+--: undesirable transfer<br/>
+
+| from\to | A   | B   | C   |
+| ---     | --- | --- | --- |
+| A 1/10  |     | -   | --  |
+| B 1/10  | -   |     | --  |
+| C 9/10  | ++  | ++  |     |
+
+To incentivize rebalancing, the node could reduce fees for channel A and B to a negative value. But this might lead to mediating transfers A&rarr;B and B&rarr;A, both of which are not intended! The problem is visible in the table above: by changing fees for outgoing channels, you can only incentivize all transfers in a single column. But all desirable transfers are in a single row, meaning the we must incentivize usage of an incoming channel.
```
Comment (palango) [inline adr/003-mediation-fees.md]: channel**s**

File: adr/003-mediation-fees.md [hunk 10]
```diff
@@ -74,6 +77,24 @@ If the node currently has a capacity of 6 and is asked to mediate a payment of 3
 
 Only mediating nodes demand mediation fees. The initiator and target could theoretically benefit if their IP was taken into account during routing, but this aspect is left out for now to avoid additional complexity.
 
+### Why apply fees for the incoming channel?
```
Comment (palango) [inline adr/003-mediation-fees.md]: 🥇 I think the example helps understand the problem!

**Comment by Dominik1999** [comment]
Thank you @karl I think this is a very good approach and I like the precise and concise description. Maybe @franzihei can even make a blogpost out of this? I think, when we follow your approach karl, we have a quite nice and unique way compared to the approaches discussed in e.g. Lightning.

**Comment by franzihei** [comment]
Regarding a blog post: Yes, that would be super nice as a piece in out "tech deep dive" series! @karlb let's sit together and briefly discuss this next week?

**Pull request merged**